### PR TITLE
Initial implementation for circuit literals

### PR DESCRIPTION
**Pull request by howardwu**
## Motivation

This PR implements circuit literals for snarkVM. This is an early prototype that will see more operator implementations for literals. In addition, an initial parser for bytecode is included to sanity check the construction for circuit literals.

## Tests

All circuit literals include tests that check for correctness and count.

## Remarks

This PR also includes:
- An update to CI which improves CI times to under 4 minutes for all jobs (except `snarkvm-gadgets`)
- Minor refactor of the `snarkvm-derives` crate into the `utilities/` folder

Backlog: This PR includes a long backlog of work from `testnet2` to prevent breakage. This size of PR is atypical.

### CODE REVIEW

File: bytecode/src/function/input.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            },
+            false => M::halt(format!("Input register {register} is not {type_}")),
+        }
+    }
+}
+
+impl<M: Memory> Operation for Input<M> {
+    type Memory = M;
+
+    /// Returns the type name as a string.
```
Comment (d0cd) [inline bytecode/src/function/input.rs]: "type name" seems to be a typo

**Review by d0cd — APPROVED** [review]

File: bytecode/src/function/input.rs (hunk 1 above)
Comment (howardwu) [inline bytecode/src/function/input.rs]:
```suggestion
    /// Returns the opcode as a string.
```

File: bytecode/src/function/output.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+/// Declares a `register` as a function output with type `annotation`.
+pub struct Output<M: Memory> {
+    argument: Argument<M::Environment>,
+}
+
+impl<M: Memory> Operation for Output<M> {
+    type Memory = M;
+
+    /// Returns the type name as a string.
```
Comment (howardwu) [inline bytecode/src/function/output.rs]:
```suggestion
    /// Returns the opcode as a string.
```

File: bytecode/src/instructions/add.rs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+/// Adds `first` with `second`, storing the outcome in `destination`.
+pub struct Add<M: Memory> {
+    operation: BinaryOperation<M::Environment>,
+}
+
+impl<M: Memory> Operation for Add<M> {
+    type Memory = M;
+
+    /// Returns the type name as a string.
```
Comment (howardwu) [inline bytecode/src/instructions/add.rs]:
```suggestion
    /// Returns the opcode as a string.
```

File: bytecode/src/instructions/store.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+// TODO: Remove store as a public facing operation.
+/// Stores `operand` into `register`, if `destination` is not already set.
+pub struct Store<M: Memory> {
+    operation: UnaryOperation<M::Environment>,
+}
+
+impl<M: Memory> Operation for Store<M> {
+    type Memory = M;
+
+    /// Returns the type name as a string.
```
Comment (howardwu) [inline bytecode/src/instructions/store.rs]:
```suggestion
    /// Returns the opcode as a string.
```

File: bytecode/src/instructions/sub.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+/// Subtracts `first` from `second`, storing the outcome in `destination`.
+pub struct Sub<M: Memory> {
+    operation: BinaryOperation<M::Environment>,
+}
+
+impl<M: Memory> Operation for Sub<M> {
+    type Memory = M;
+
+    /// Returns the type name as a string.
```
Comment (howardwu) [inline bytecode/src/instructions/sub.rs]:
```suggestion
    /// Returns the opcode as a string.
```

**Pull request merged**